File: Greedy Algo/knapsack.java
/* this is a greedy type algo to find maximize profit it takes weight and value and find maximum profit */ 

import java.util.Scanner;

class knapsack
{
	public static double calRatio(int a,int b)
	{
		double res;
		res = (double)(b/a);
		return res;
	}

	public static double knapalgo(int wt[],int val[],int cap)
	{
		double valwt[] = new double[wt.length];
		double totalVal = 0;
		
		for(int i=0;i<wt.length;i++)
		{
			valwt[i] = calRatio(wt[i],val[i]);
		}
		// for(int i=0;i<wt.length;i++)
		// {
		// 	System.out.println(valwt[i]);
		// }
		while(cap>0)
		{
			double max = valwt[0];
			int index = 0;
			for(int i=1;i<valwt.length;i++)
			{
				if(max<valwt[i])
				{
					max = valwt[i];
					index = i;
				}
			}
			valwt[index] = -1;
			if(cap-wt[index] >= 0 )
			{
				cap = cap-wt[index];
				totalVal = totalVal+val[index];
			}
			else
			{
				double fraction = ((double)cap/(double)wt[index]);
				totalVal += (val[index]*fraction);
				cap = (int)(cap - (wt[index]*fraction));
				break;
			}
		}

		
		return totalVal;
	}

	public static void main(String[] args) 
	{
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter Size : ");
		int n = sc.nextInt();

		int wt[] = new int[n];
		int val[] = new int[n];

		System.out.println("Enter Weight Elements : ");
		for(int i=0;i<n;i++)
		{
			wt[i] = sc.nextInt();
		}

		System.out.println("Enter Value Elements : ");
		for(int i=0;i<n;i++)
		{
			val[i] = sc.nextInt();
		}

		System.out.println("Enter Capacity : ");
		int cap = sc.nextInt();

		double res = knapalgo(wt,val,cap);
		System.out.println(res);
		
	}
}

File: DYNAMIC PROGRAMMING/knapSack.c
#include<stdio.h>

int max(int a,int b)
{
	if(a>b)
		return a;
	else
		return b;
}

void knap(int item,int cap,int p[],int w[])
{
	int knap[item+1][cap+1];
	for(int i=0;i<=item;i++)
	{
		for(int j=0;j<=cap;j++)
		{
			if(i==0 || j==0)
				knap[i][j] = 0;
			else if(w[i-1] <= j)
				knap[i][j] = max(knap[i-1][j-w[i-1]] + p[i-1] , knap[i-1][j]);
			else
				knap[i][j] = knap[i-1][j];
		}
	}
	printf("%d\n",knap[item][cap] );

}

int main()
{
	int p[] = {10,12,28};
	int w[] = {1,2,4};
	int c = 6;
	knap(3,c,p,w);
}